### PR TITLE
kubectl run set cpu/memory request/limit

### PR DESCRIPTION
**Pull request by derekwaynecarr**
Let's a user specify resource requirements when performing a kubectl run command.

Fixes https://github.com/kubernetes/kubernetes/issues/8328

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
@kubernetes/kubectl folks will need to review this.

File: pkg/kubectl/cmd/run.go [hunk 1]
```diff
@@ -85,6 +85,10 @@ func NewCmdRun(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *c
 	cmd.Flags().Bool("attach", false, "If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...' were called.  Default false, unless '-i/--interactive' is set, in which case the default is true.")
 	cmd.Flags().String("restart", "Always", "The restart policy for this Pod.  Legal values [Always, OnFailure, Never].  If set to 'Always' a replication controller is created for this pod, if set to OnFailure or Never, only the Pod is created and --replicas must be 1.  Default 'Always'")
 	cmd.Flags().Bool("command", false, "If true and extra arguments are present, use them as the 'command' field in the container, rather than the 'args' field which is the default.")
+	cmd.Flags().String("cpu-request", "", "The CPU request for the container expressed in millicores.")
```
Comment (kargakis) [inline pkg/kubectl/cmd/run.go]: We might want these grouped like https://github.com/kubernetes/kubernetes/blob/d23d083439d4b936ceaa82d562fc9880906412c4/pkg/kubectl/cmd/util/printing.go#L30

Comment (derekwaynecarr) [inline pkg/kubectl/cmd/run.go]: I think its good to move if/when we have two commands that take them as input.  So for now, I would prefer to keep it in run.

**Comment by smarterclayton** [comment]
What are we going to do when we have more resource types?  Lots of flags?

**Comment by smarterclayton** [comment]
Could we instead add --limit=cpu=1 or something more flexible to future expression?

**Comment by derekwaynecarr** [comment]
@smarterclayton - I guess this pattern would introduce a number of flags for other resource constraints in the future.  I am not sure how many pod level request/limits we imagine in the future.  There was some discussion on bandwidth, but maybe its not too bad.

If we go with a single flag style we need a tokenizer.  Is there a standard token flag we have been using?  Or would you just express the flag multiple times?

```
--limit=cpu=1 --limit=memory=512Mi --request=cpu=500m --request=memory=100Mi
```

**Comment by smarterclayton** [comment]
Multiple times probably, but almost certainly a comma won't be in the
allowed characters list for a resource name, so we could potentially have
the simpler `--limit=cpu=1,memory=512Mi`

On Tue, Sep 8, 2015 at 10:39 AM, Derek Carr <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> - I guess this
> pattern would introduce a number of flags for other resource constraints in
> the future. I am not sure how many pod level request/limits we imagine in
> the future. There was some discussion on bandwidth, but maybe its not too
> bad.
>
> If we go with a single flag style we need a tokenizer. Is there a standard
> token flag we have been using? Or would you just express the flag multiple
> times?
>
> --limit=cpu=1 --limit=memory=512Mi --request=cpu=500m --request=memory=100Mi
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13460#issuecomment-138586182>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by derekwaynecarr** [comment]
I will update this PR to go single flag style ```limit``` and ```request``` and use the comma as a token.

Example style will become:

```kubectl run nginx --image=nginx --replicas=3 --limit=cpu=100m,memory=512Mi --request=cpu=10m,memory=256Mi```

**Comment by derekwaynecarr** [comment]
@smarterclayton - updated PR to use ```limits=``` and ```requests=``` syntax.

This should be good to review.

**Comment by smarterclayton** [comment]
No significant issues.  Would we expose a limits alteration command with similar syntax?  If so the helpers for creating the pod would probably be good to make public and move out of run.go.  That's not important enough to do here, although I would like this usable with oc new-app

**Comment by bgrant0607** [comment]
Would be great to get this in 1.1.

File: pkg/kubectl/run.go [hunk 2]
```diff
@@ -42,9 +43,44 @@ func (BasicReplicationController) ParamNames() []GeneratorParam {
...
 		{"args", false},
 		{"env", false},
+		{"requests", false},
+		{"limits", false},
 	}
 }
 
+func handleResourceRequirement(params map[string]string) (api.ResourceRequirements, error) {
+	result := api.ResourceRequirements{}
+	for _, paramName := range []string{"requests", "limits"} {
```
Comment (brendandburns) [inline pkg/kubectl/run.go]: This code is awkward.

Why not:

```go
func populateResourceList(list *api.ResourceList, spec string) {
   ...
}

func handleResourceRequirements(...) {
    populateResourceList(&result.Limits, params["limits"])
    populateResourceList(&result.Requests, params["requests"])
}

**Comment by brendandburns** [comment]
one comment on code factoring, and it needs a rebase.

**Comment by derekwaynecarr** [comment]
@brendandburns - refactored code

@smarterclayton - made the parser for limits and cpu public so it can be used in ```oc``` but its still in ```run.go``` for now.

this should be good to go for now.

**Comment by smarterclayton** [comment]
LGTM

**Pull request merged**